llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4728,6 +4740,47 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   // FIXME: look for a smaller MaxVF that does divide TC rather than masking.
   if (Legal->prepareToFoldTailByMasking()) {
     CanFoldTailByMasking = true;
+    if (getTailFoldingStyle() == TailFoldingStyle::DataWithEVL) {
+      assert(!ChosenTailFoldingStyle &&
```

Comment:
Better to move this to its own helper (e.g. `selectTailFoldingStyle`), have to called before computeMaxVF and move all logic related to selecting the tail-folding style there, so that `getTailFoldingStyle()` is simplify a return of the selected style? Would be good to do this refactoring in a separate PR